### ISSUE TITLE
add json to expect_request

### ISSUE THREAD

**Issue by fredwangwang** [csernazs/pytest-httpserver#41]
right now `data` can only match exact text, which is less useful for testing apis that takes json payload.

expect_request should be able to take `data` or `json` (mutually exclusive) and handle the match correspondingly.

btw, thanks for the great project! Saves wring a bunch of boilerplates code!

**Comment by csernazs** [comment]
This is something I wanted to implement. I think I wanted to implement it in a wrong way: dumping the json provided and then comparing it with the raw request body, which is troublesome due to the various ways of json encoding (especially the indentation).

So this could be implemented in the opposite: loading the raw request with json.loads() then comparing it to the provided python dict/whatever.

The only issue I see is that what happens when the request is not a json and json.loads() raises an error? All we can do is to log it, then respond with http status 500, as the server is running in a different thread.
But that is also an issue if the content cannot be decoded to string which is handled correctly.

**Comment by csernazs** [comment]
I've made a PoC for the json matcher, if time allows, could you please look at it? Is this something you wanted?

**Comment by fredwangwang** [comment]
You get it implemented fast! But I think there is a bug in the code: `is_json` is an attribute and is not callable. 
Here is the diff:

```
diff --git a/pytest_httpserver/httpserver.py b/pytest_httpserver/httpserver.py
index 0651420..3d7e702 100644
--- a/pytest_httpserver/httpserver.py
+++ b/pytest_httpserver/httpserver.py
@@ -345,7 +345,7 @@ class RequestMatcher:
             return self.uri == URI_DEFAULT or path == self.uri
 
     def match_json(self, request: Request) -> bool:
-        return self.json is None or request.is_json() and request.get_json() == self.json
+        return self.json is None or request.is_json and request.get_json() == self.json
 
     def difference(self, request: Request) -> list:
         """
diff --git a/tests/test_json_matcher.py b/tests/test_json_matcher.py
index 8b09b74..e08d124 100644
--- a/tests/test_json_matcher.py
+++ b/tests/test_json_matcher.py
@@ -7,4 +7,4 @@ import requests
 def test_json_matcher(httpserver: HTTPServer):
     httpserver.expect_request("/foo", json={"foo": "bar"}).respond_with_data("Hello world!")
     assert requests.get(httpserver.url_for("/foo")).status_code == 500
-    assert requests.get(httpserver.url_for("/foo"), json={"foo": "bar"}).status_code == 500
+    assert requests.get(httpserver.url_for("/foo"), json={"foo": "bar"}).status_code == 200

```

**Comment by csernazs** [comment]
Argh, I copy&pasted the http status 500 in the test by mistake and I believed that the test passed.. Good catch, thank you!

**Comment by csernazs** [comment]
One question: is it ok to have the "is_json()" called? It checks that the content-type is specified correctly.

So currently I can see two possible implementations:

1. do it as the current implementation (with your fix), it checks the content-type header and it relies on werkzeug which is good as they probably do it better than me
2. just load the json and compare it, eg no werkzeug will be called and the content-type is not checked at all. If it is desired, it can be added to the headers parameter.

For me the 2nd is a bit more "flexible" but on the other hand the 1st is very standard and it could "educate" the people to implement the REST APIs correctly.

**Comment by csernazs** [comment]
I tend to implement the second case (patch pushed, see above) as the user can still add the content-length checking if it is desired but removing the content-length checking from the 1st case would be an API break.

Any comments on this? If it suits you, I would move toward with code clean-up, etc and integration.

**Comment by fredwangwang** [comment]
Looks like should be fine with me, thanks for checking!

**Comment by csernazs** [comment]
Hi @fredwangwang,

I've just released pytest-httpserver 0.3.6 to pypi with both of your requests implemented (and also #44).
JSON matching even has its own howto page here: https://pytest-httpserver.readthedocs.io/en/latest/howto.html#json-matching

Enjoy and thanks!

Zsolt

**Comment by fredwangwang** [comment]
Nice work, thank you @csernazs !
